### PR TITLE
Update leaf syntax to leaf 4

### PR DESCRIPTION
**Pull request by dannflor**
### Changes
Overhauled the leaf syntax to match the new leaf 4 grammar
https://docs.vapor.codes/leaf/overview

### Checklist
- [x] Added markup tests, or they don't apply here because...
- [ ] Updated the changelog at `CHANGES.md`

### CODE REVIEW

File: src/languages/leaf.js [hunk 1]
```diff
@@ -1,23 +1,23 @@
...
-Author: Hale Chan <halechan@qq.com>
-Description: Based on the Leaf reference from https://vapor.github.io/documentation/guide/leaf.html.
+Author: Samuel Bishop <sambobjr@gmail.com>
+Description: A Swift-based templating language created for the Vapor project.
+Website: https://docs.vapor.codes/leaf/overview
+Category: template
 */
 
-export default function(hljs) {
+export default function leaf(hljs) {
```
Comment (joshgoebel) [inline src/languages/leaf.js:9]: Revert please. This is purposely unnamed.

File: src/languages/leaf.js [hunk 2]
```diff
@@ -26,22 +26,57 @@ export default function(hljs) {
...
+                className: 'keyword',
+                begin: 'true|false|in',
+              },
               {
                 className: 'variable',
                 begin: '[A-Za-z_][A-Za-z_0-9]*'
-              }
+              },
+              {
+                className: 'operator',
```
Comment (joshgoebel) [inline src/languages/leaf.js:47]: Operators can only exist within params?

Comment (dannflor) [reply src/languages/leaf.js:47]: Correct. Leaf is a templating language so there's no use for them outside of a pair of parentheses

File: src/languages/leaf.js [hunk 3]
```diff
@@ -26,22 +26,57 @@ export default function(hljs) {
...
             endsParent: true,
+            relevance: 7,
             contains: [
               {
                 className: 'string',
                 begin: '"',
                 end: '"'
               },
+              {
+                className: 'keyword',
```
Comment (joshgoebel) [inline src/languages/leaf.js:39]: Ditto, are these only valid within params?

Comment (dannflor) [reply src/languages/leaf.js:39]: See above

**Comment by joshgoebel** [comment]
Thanks, but it looks like this whole grammar is super old and using very old conventions - we have nicer ways to do a lot of this now.  I'll try to pull this down and make a pass at cleaning it up then toss it back.

Is there a list of official helpers/keywords in Leaf (loop, for, etc)?  If so we should build them in so they can be highlighted as `built-in` perhaps...

**Comment by dannflor** [comment]
https://docs.vapor.codes/leaf/overview is basically all you have to work off of. Anyone can define any tag though, so it doesn't really make sense to distinguish between built-in and user defined. The only built in tag that has any different syntax from an arbitrarily user defined one is `for` requiring an `in` in the parameter body. But you could also mark every one of the tags that is on that page as built-in.

**Comment by joshgoebel** [comment]
Screen Shot 2023-05-25 at 10 00 55 PM

Pretty sure if/else/etc should all match though... ie, they are a single scope...

**Comment by dannflor** [comment]
Sounds good to me. We've been living without syntax highlighting on those specific keywords for a couple of years now so I don't know if there are any strong opinions lol

**Comment by dannflor** [comment]
I updated the expected markup to match what we want to generate. There's only one thing missing, which is tag functions can nest inside each other (nested functions omit the `#` at the beginning). I believe they can nest to an arbitrary depth, but I know regular expressions can't do that (though maybe you have some way around that). I'm not sure how to implement that in your current system for defining the grammar

**Comment by joshgoebel** [comment]
> which is tag functions can nest inside each other (nested functions omit the # at the beginning)

Give me a few actual examples please.

**Comment by dannflor** [comment]
Sorry, I put it in the markup test. It's basically something like:
`#if(count(arbitaryFunction(variable))) == 0):`

**Comment by joshgoebel** [comment]
Regex doesn't support infinite nesting, but the parser does... for something like one rule immediately inside another we have a shortcut "self"... for less-direct looping you have to create the objects separate then link them together as the grammar is executed. IE:

- a
  - b
    - c

Where a contains b which contains c, etc... but there is nothing to stop `C` from containing `A`... and it can go as deep as necessary....

**Comment by dannflor** [comment]
Hey, it works! LGTM

**Comment by dannflor** [comment]
Is there something remaining to be done here holding up merging?

**Comment by joshgoebel** [comment]
Yeah, just need to update the changelog :)

**Comment by dannflor** [comment]
Done

**Review by joshgoebel — APPROVED** [review]

**Comment by joshgoebel** [comment]
Thanks!

**Pull request merged**